### PR TITLE
BigQuery Additional e2e Test.

### PR DESCRIPTION
**Pull request by Praveena2607**

### CODE REVIEW

File: src/e2e-test/features/bigquery/source/BigQuerySourceError.feature [hunk 1]
```diff
@@ -55,3 +55,24 @@ Feature: BigQuery source - Validate BigQuery source plugin error scenarios
     Then Enter BigQuery source property table name
     Then Enter BigQuery property temporary bucket name "bqInvalidTemporaryBucket"
     Then Verify the BigQuery validation error message for invalid property "bucket"
+
+  @BQ_SOURCE_TEST
+  Scenario Outline:Verify if invalid Partition Start date and Partition end Date format at design time or check pipeline fails while preview
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQuerySourceError.feature]: Change the scenario title as per the scenario, here we are not checking at preview.

File: src/e2e-test/resources/pluginParameters.properties [hunk 2]
```diff
@@ -1,6 +1,16 @@
...
+partitionTo=2024-12-04
+filter=Id=20
+OutputSchemamacroinput={ "type": "record", "name": "text", "fields": [ \
+  { "name": "ProjectId", "type": "string" }, { "name": "Dataset", "type": "string" } ] }
 datasetprojectId=cdf-athena
 dataset=testbq_bqmt
+bqIncorrectFormatStartDate=dfgfdsdfd
+bqIncorrectFormatEndDate=oioiojojojjoj
+bqIncorrectReferenceName=#$%^
+bqIncorrectFilter=%%%%
```
Comment (itsmekumari) [inline src/e2e-test/resources/pluginParameters.properties]: Remove this unused parameter value.

File: src/e2e-test/resources/pluginParameters.properties [hunk 3]
```diff
@@ -1,6 +1,16 @@
...
+#partitionFrom={"2024-11-25"}
```
Comment (itsmekumari) [inline src/e2e-test/resources/pluginParameters.properties]: Remove the unused parameter value.

File: src/e2e-test/resources/pluginParameters.properties [hunk 4]
```diff
@@ -1,6 +1,16 @@
 projectId=cdf-athena
+partitionFrom=2024-12-03
```
Comment (itsmekumari) [inline src/e2e-test/resources/pluginParameters.properties]: Remove from here and add all the BigQuery related changes in BigQuery properties section only

File: src/e2e-test/resources/pluginParameters.properties [hunk 5]
```diff
@@ -195,13 +208,15 @@ bqPartitionField=Month_of_Joining
...
+bqStartDate=2024-12-03
+bqEndDate=2024-12-04
 bqRangeInterval=2
 bqChunkSize=262144
 bqInvalidChunkSize=26214
 bqInvalidSinkDataset=$#%$
 bqInvalidSinkTable=(*^*&*
 bqInvalidTemporaryBucket=$#%$
-bqInvalidRefName=invalidRef&^*&&*
+bqInvalidRefName=#$%^
```
Comment (itsmekumari) [inline src/e2e-test/resources/pluginParameters.properties]: Why are we changing the value here?

File: src/e2e-test/java/io/cdap/plugin/utils/CdfPluginPropertyLocator.java [hunk 6]
```diff
@@ -46,35 +50,43 @@ public enum CdfPluginPropertyLocator {
   private static final Map<String, CdfPluginPropertyLocator> CDF_PLUGIN_PROPERTY_MAPPING;
   static {
     CDF_PLUGIN_PROPERTY_MAPPING = new ImmutableMap.Builder<String, CdfPluginPropertyLocator>()
-      .put("projectId", CdfPluginPropertyLocator.PROJECT_ID)
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/utils/CdfPluginPropertyLocator.java:49]: Why this changes made in this file? Revert it

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 7]
```diff
@@ -179,6 +188,18 @@ public void validateTheCmekKeyOfTargetBigQueryTableIfCmekIsEnabled(String cmek)
   public void enterBigQueryPropertyAsMacroArgument(String pluginProperty, String macroArgument) {
     enterPropertyAsMacroArgument(pluginProperty, macroArgument);
   }
+  @Then("Enter BigQuery source property output schema {string} as macro argument {string}")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:188]: We already have an existing step to select macro action for output schema, use that. Always try to use the existing steps if we already have. Do not create the step definition again.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 8]
```diff
@@ -18,23 +18,26 @@
 import io.cdap.e2e.pages.actions.CdfBigQueryPropertiesActions;
 import io.cdap.e2e.pages.actions.CdfStudioActions;
 import io.cdap.e2e.pages.locators.CdfStudioLocators;
-import io.cdap.e2e.utils.BigQueryClient;
-import io.cdap.e2e.utils.ConstantsUtil;
-import io.cdap.e2e.utils.ElementHelper;
-import io.cdap.e2e.utils.PluginPropertyUtils;
-import io.cdap.e2e.utils.SeleniumHelper;
+import io.cdap.e2e.utils.*;
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:24]: Make sure existing imports are not removed, and since * is added you will get the check style issue. Please add in the order it was and check once if any check style issues occurs resolve it.

File: src/e2e-test/features/bigquery/source/BigQueryToGCS_WithMacro.feature [hunk 9]
```diff
@@ -69,3 +69,223 @@ Feature: BigQuery source - Verification of BigQuery to GCS successful data trans
...
+    Then Open BigQuery source properties
+    Then Enter BigQuery property reference name
+    Then Enter BigQuery property "projectId" as macro argument "bqProjectId"
+    Then Enter BigQuery property "datasetProjectId" as macro argument "bqDatasetProjectId"
+    Then Enter BigQuery property "serviceAccountType" as macro argument "serviceAccountType"
+    Then Enter BigQuery property "serviceAccountFilePath" as macro argument "serviceAccount"
+    Then Enter BigQuery property "serviceAccountJSON" as macro argument "serviceAccount"
+    Then Enter BigQuery property "dataset" as macro argument "bqDataset"
+    Then Enter BigQuery property "table" as macro argument "bqSourceTable"
+    Then Enter BigQuery source property output schema "outputSchema" as macro argument "bqOutputSchema"
```
Comment (itsmekumari) [inline src/e2e-test/features/bigquery/source/BigQueryToGCS_WithMacro.feature:234]: Use the existing step definition for selecting macro action for output schema as mentioned above.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 10]
```diff
@@ -66,6 +69,12 @@ public void enterBigQueryPropertyDataset(String dataset) {
     CdfBigQueryPropertiesActions.enterBigQueryDataset(PluginPropertyUtils.pluginProp(dataset));
   }
 
+  @Then("Enter BigQuery property partitionFrom {string}")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java]: Remove this step as it is not used and needed.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 11]
```diff
@@ -110,13 +119,13 @@ public void validateRecordsTransferredToTargetTableIsEqualToNumberOfRecordsFromS
...
   public void validatePartitionDateInOutputPartitionedTable()
-    throws IOException, InterruptedException {
+          throws IOException, InterruptedException {
     Optional<String> result = BigQueryClient
-      .getSoleQueryResult("SELECT distinct  _PARTITIONDATE as pt FROM `" +
-                            (PluginPropertyUtils.pluginProp("projectId")) + "." +
-                            (PluginPropertyUtils.pluginProp("dataset")) + "." +
-                            TestSetupHooks.bqTargetTable +
-                            "` WHERE _PARTITION_LOAD_TIME IS Not NULL ORDER BY _PARTITIONDATE DESC ");
+            .getSoleQueryResult("SELECT distinct  _PARTITIONDATE as pt FROM `" +
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:121]: Why are these changes are made, revert all the unnecessary changes made.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 12]
```diff
@@ -224,22 +245,32 @@ public void enterBigQuerySinkPropertyTemporaryBucketName(String temporaryBucket)
     CdfBigQueryPropertiesActions.enterTemporaryBucketName(PluginPropertyUtils.pluginProp(temporaryBucket));
   }
 
+  @Then("Enter BigQuery property reference name {string}")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java]: Use the existing step to enter the value in any property, which I shared earlier from framework. Remove this step definition.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 13]
```diff
@@ -248,16 +279,60 @@ public void verifyTheBigQueryValidationErrorMessageForInvalidProperty(String pro
     Assert.assertEquals(expectedColor, actualColor);
   }
 
+  @Then("Verify the BigQuery validation error message for reference name {string}")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:274]: Why are we duplicating the step definition. I already informed and mentioned long back to go through the existing tests and use the existing steps. Here as well, no need to create this step definition again. You can use the existing  in-line error message verify step. And also I don't think you are using this in your scenario.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java [hunk 14]
```diff
@@ -248,16 +279,60 @@ public void verifyTheBigQueryValidationErrorMessageForInvalidProperty(String pro
...
-      ("dataset")) + "." + TestSetupHooks.bqTargetTable + "` ");
+            (PluginPropertyUtils.pluginProp("projectId"))
+            + "." + (PluginPropertyUtils.pluginProp
+            ("dataset")) + "." + TestSetupHooks.bqTargetTable + "` ");
     int count = result.map(Integer::parseInt).orElse(0);
     BeforeActions.scenario.write("Number of records transferred from source table to target table:" + count);
     Assert.assertEquals(count, countRecordsTarget);
   }
+
+  @Then("Enter BigQuery property output schema {string} as macro argument {string}")
```
Comment (itsmekumari) [inline src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java]: This is repeated step, and also as mentioned remove it as we already have an existing step to select macro action for output schema, use that. Always try to use the existing steps if we already have. Do not create the step definition again.

File: src/e2e-test/resources/pluginParameters.properties (hunk 5 above)
Comment (Praveena2607) [reply src/e2e-test/resources/pluginParameters.properties]: Changed as it is.

File: src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java (hunk 11 above)
Comment (Praveena2607) [reply src/e2e-test/java/io/cdap/plugin/bigquery/stepsdesign/BigQueryBase.java:121]: I didn't touch this line.
